File: src/main/java/com/example/visainterview/service/TransactionService.java
package com.example.visainterview.service;

import com.example.visainterview.dto.CreateTransactionRequest;
import com.example.visainterview.dto.TransactionResponse;
import com.example.visainterview.entity.Account;
import com.example.visainterview.entity.Transaction;
import com.example.visainterview.entity.TransactionType;
import com.example.visainterview.exception.AccountNotFoundException;
import com.example.visainterview.exception.InsufficientBalanceException;
import com.example.visainterview.exception.InvalidTransactionTypeException;
import com.example.visainterview.repository.AccountRepository;
import com.example.visainterview.repository.TransactionRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class TransactionService {

    private final TransactionRepository transactionRepository;
    private final AccountRepository accountRepository;

    @Transactional
    public TransactionResponse createTransaction(CreateTransactionRequest request) {
        Optional<Transaction> existingTransaction = transactionRepository
                .findByIdempotencyKey(request.getIdempotencyKey());
        if (existingTransaction.isPresent()) {
            return mapToResponse(existingTransaction.get());
        }

        TransactionType transactionType;
        try {
            transactionType = TransactionType.fromCode(request.getTransactionType());
        } catch (IllegalArgumentException e) {
            throw new InvalidTransactionTypeException("Invalid transaction type: " + request.getTransactionType());
        }

        Account account = accountRepository.findById(request.getAccountId())
                .orElseThrow(() -> new AccountNotFoundException("Account not found"));

        if (transactionType.isDebit()) {
            if (account.getBalance().compareTo(request.getAmount()) < 0) {
                throw new InsufficientBalanceException("Insufficient balance");
            }
            account.setBalance(account.getBalance().subtract(request.getAmount()));
        } else if (transactionType.isCredit()) {
            account.setBalance(account.getBalance().add(request.getAmount()));
        }

        accountRepository.save(account);

        Transaction transaction = Transaction.builder()
                .account(account)
                .transactionType(transactionType)
                .amount(request.getAmount())
                .idempotencyKey(request.getIdempotencyKey())
                .build();
        transaction = transactionRepository.save(transaction);

        return mapToResponse(transaction);
    }

    private TransactionResponse mapToResponse(Transaction transaction) {
        return TransactionResponse.builder()
                .transactionId(transaction.getId())
                .accountId(transaction.getAccount().getId())
                .transactionType(transaction.getTransactionType().getCode())
                .amount(transaction.getAmount())
                .transactionDate(transaction.getCreatedAt())
                .build();
    }
}

File: golang/service/transaction_service.go
package service

import (
	"context"
	"errors"

	"github.com/visainterview/financial-transaction-system/dto"
	"github.com/visainterview/financial-transaction-system/middleware"
	"github.com/visainterview/financial-transaction-system/models"
	"github.com/visainterview/financial-transaction-system/repository"
	"github.com/visainterview/financial-transaction-system/utils"
	"gorm.io/gorm"
)

type TransactionService interface {
	CreateTransaction(ctx context.Context, req dto.CreateTransactionRequest) (*dto.TransactionResponse, error)
}

type transactionService struct {
	transactionRepo repository.TransactionRepository
	accountRepo     repository.AccountRepository
}

func NewTransactionService(transactionRepo repository.TransactionRepository, accountRepo repository.AccountRepository) TransactionService {
	return &transactionService{
		transactionRepo: transactionRepo,
		accountRepo:     accountRepo,
	}
}

func (s *transactionService) CreateTransaction(ctx context.Context, req dto.CreateTransactionRequest) (*dto.TransactionResponse, error) {
	transactionType := models.TransactionType(req.TransactionType)
	if !transactionType.IsValid() {
		return nil, utils.ErrInvalidTransactionType
	}

	existingTxn, err := s.transactionRepo.FindByIdempotencyKey(ctx, req.IdempotencyKey)
	if err == nil && existingTxn != nil {
		return &dto.TransactionResponse{
			TransactionID:   existingTxn.ID,
			AccountID:       existingTxn.AccountID,
			TransactionType: int(existingTxn.TransactionType),
			Amount:          existingTxn.Amount,
			TransactionDate: existingTxn.CreatedAt,
		}, utils.ErrIdempotencyKeyExists
	}

	var transaction *models.Transaction
	const maxRetries = 2

	for attempt := 0; attempt <= maxRetries; attempt++ {
		account, err := s.accountRepo.FindByID(ctx, req.AccountID)
		if err != nil {
			if errors.Is(err, gorm.ErrRecordNotFound) {
				return nil, utils.ErrAccountNotFound
			}
			return nil, err
		}

		newBalance := account.Balance
		if transactionType.IsDebit() {
			if account.Balance < req.Amount {
				return nil, utils.ErrInsufficientBalance
			}
			newBalance -= req.Amount
		} else if transactionType.IsCredit() {
			newBalance += req.Amount
		}

		if transaction == nil {
			transaction = &models.Transaction{
				AccountID:       req.AccountID,
				TransactionType: transactionType,
				Amount:          req.Amount,
				IdempotencyKey:  req.IdempotencyKey,
			}

			if err := s.transactionRepo.Create(ctx, transaction); err != nil {
				return nil, err
			}
		}

		err = s.accountRepo.UpdateBalance(ctx, account, newBalance)
		if err == nil {
			txTypeStr := transactionType.String()
			middleware.RecordTransaction(txTypeStr, "success", req.Amount)

			return &dto.TransactionResponse{
				TransactionID:   transaction.ID,
				AccountID:       transaction.AccountID,
				TransactionType: int(transaction.TransactionType),
				Amount:          transaction.Amount,
				TransactionDate: transaction.CreatedAt,
			}, nil
		}

		if !errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, err
		}

		middleware.RecordOptimisticLockRetry()

		if attempt == maxRetries {
			return nil, utils.ErrOptimisticLockFailed
		}
	}

	return nil, utils.ErrOptimisticLockFailed
}
